runtime/near-vm-logic/src/logic.rs

```diff
@@ -639,6 +640,29 @@ impl<'a> VMLogic<'a> {
         Ok(self.context.block_height)
     }
 
+    /// Gets a block given a height (up to 256)
+    ///
+    /// # Cost
+    ///
+    /// `base`
+    #[cfg(feature = "protocol_feature_get_block_by_hash")]
+    pub fn get_block_hash_by_height(
+        &mut self,
+        height_number: BlockHeight,
+        register_id: u64,
+    ) -> Result<u64> {
+        self.gas_counter.pay_base(storage_block_hash_base)?;
+        let (iterations, block_hash) = self.ext.get_block_hash_by_height(height_number);
+        self.gas_counter.pay_per(storage_iter_next_block_hash, iterations as _)?;
```

Comment:
In the spirit of charging gas before doing the work, the per-hash payment should actually happen in every step of iteration.
(See https://near.github.io/nearcore/architecture/gas/gas_profile.html for the reasoning)

Reply:
this makes sense